### PR TITLE
[diffusion] feat: enable torch compile to eliminate GPU bubble

### PR DESCRIPTION
**Pull request by AichenF**
@jianyingzhu 

## Motivation

https://github.com/sgl-project/sglang/issues/12988

## Modifications

Enable torch.compile to eliminate GPU bubbles. We currently use the _max-autotune-no-cudagraphs_ compile mode, since _max-autotune_ introduces significant delays during compilation.

## Benchmarking and Profiling

on H100

|  | without torch.compile | with torch.compile Max-autotune | with torch.compile Max-autotune-no-cudagraphs |
| ---------------------------------------- | ------------------------ | ------------ | ----------------------------- |
| Time per denoising step in torch profile | 502ms | 190ms | 416ms |

without torch.compile
```
sglang generate --model-path=Qwen/Qwen-Image --prompt='a curious raccoon' --width=720 --height=720 --save-output
```

image

with torch.compile (max-autotune-no-cudagraphs)
```
sglang generate --model-path=Qwen/Qwen-Image --prompt='a curious raccoon' --width=720 --height=720 --save-output --enable-torch-compile
```

image

with torch.compile (max_autotune)

image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 1]
```diff
@@ -197,10 +206,22 @@ def _prepare_denoising_loop(self, batch: Req, server_args: ServerArgs):
...
+                    _inductor_cfg.reorder_for_compute_comm_overlap = True
+                except Exception:
+                    pass
+                mode = os.environ.get(
+                    "SGLANG_TORCH_COMPILE_MODE", "max-autotune-no-cudagraphs"
+                )
+                compiled_forward = torch.compile(
+                    getattr(self.transformer, "forward"), mode=mode
                 )
+                setattr(self.transformer, "forward", compiled_forward)
```
[the remark this answers is not in the document]
Comment (jianyingzhu) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Same as before, in `_prepare_denoising_loop`, torch.compile is only enabled when `server_args.model_loaded["transformer"]` equals to False, while  The `server_args.model_loaded["transformer"]` parameter cannot be controlled by the user when launching the server and is instead set to `False` by default in the ServerArgs class:
```
    model_loaded: dict[str, bool] = field(
        default_factory=lambda: {
            "transformer": True,
            "vae": True,
        }
    )
```
So in `_prepare_denoising_loop`, the code inside the if-branch (load transformer block and substitute its forward with torch compiled forward), will not be executed.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 2]
```diff
@@ -112,17 +112,26 @@ def __init__(
...
                 )
-                if transformer_2 is not None
-                else None
-            )
+                setattr(self.transformer, "forward", compiled_forward)
+            if transformer_2 is not None and hasattr(self.transformer_2, "forward"):
+                compiled_forward_2 = torch.compile(
+                    getattr(self.transformer_2, "forward"), mode=mode
+                )
+                setattr(self.transformer_2, "forward", compiled_forward_2)
```
[the remark this answers is not in the document]
Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: We have deleted the duplicated torch.compile in `composed_pipeline_base.py`. In `_prepare_denoising_loop`, torch.compile is only enabled when `server_args.model_loaded["transformer"]` equals to False, while  `server_args.model_loaded["transformer"]` is always true, set in the ServerArgs class:
```
    model_loaded: dict[str, bool] = field(
        default_factory=lambda: {
            "transformer": True,
            "vae": True,
        }
    )
```
So in `_prepare_denoising_loop`, the code inside the if-branch (load transformer block and substitute its forward with torch compiled forward), will not be executed.

File: python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py [hunk 3]
```diff
@@ -105,8 +105,25 @@ def post_init(self) -> None:
...
+            transformer = self.modules.get("transformer")
+            if transformer is not None and hasattr(transformer, "forward"):
+                compiled_forward = torch.compile(
+                    getattr(transformer, "forward"), mode=mode
+                )
+                setattr(transformer, "forward", compiled_forward)
+                self.modules["transformer"] = transformer
+                logger.info(
+                    "Torch Compile enabled for DiT (mode=%s, forward-wrapped)", mode
+                )
```
[the remark this answers is not in the document]
Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py]: Same as before, in `_prepare_denoising_loop`, torch.compile is only enabled when `server_args.model_loaded["transformer"]` equals to False, while  The `server_args.model_loaded["transformer"]` parameter cannot be controlled by the user when launching the server and is instead set to `False` by default in the ServerArgs class:
```
    model_loaded: dict[str, bool] = field(
        default_factory=lambda: {
            "transformer": True,
            "vae": True,
        }
    )
```
So in `_prepare_denoising_loop`, the code inside the if-branch (load transformer block and substitute its forward with torch compiled forward), will not be executed.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 4]
```diff
@@ -116,17 +116,26 @@ def __init__(
...
+            except Exception:
+                pass
+            mode = os.environ.get(
+                "SGLANG_TORCH_COMPILE_MODE", "max-autotune-no-cudagraphs"
             )
-            self.transformer_2 = (
-                torch.compile(
-                    self.transformer_2, mode="max-autotune", fullgraph=full_graph
+            # Only compile when it's a module-like object (has forward). Wrap forward like xDiT.
+            if hasattr(self.transformer, "forward"):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: nit: in what circumstance, the transformer does not have a `forward`

Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: We should have something like `def torch_compile_module()` to help with the _inductor_cfg importing and actual torch_compile logic

Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Also, do we need to cover `denoising_dmd`?

Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: This is only for safety checks and compiles only the forward pass instead of the entire transformer.

Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: > Also, do we need to cover `denoising_dmd`?

haven’t tested denoising_dmd yet. Which model will use this stage？

Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: > We should have something like `def torch_compile_module()` to help with the _inductor_cfg importing and actual torch_compile logic

ok

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 5]
```diff
@@ -116,17 +116,26 @@ def __init__(
...
         # torch compile
         if self.server_args.enable_torch_compile:
-            full_graph = False
-            self.transformer = torch.compile(
-                self.transformer, mode="max-autotune", fullgraph=full_graph
+            try:
+                import torch._inductor.config as _inductor_cfg
+
+                _inductor_cfg.reorder_for_compute_comm_overlap = True
+            except Exception:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Try-Catch `ImportError` is better?

File: python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py [hunk 6]
```diff
@@ -110,9 +110,7 @@ def post_init(self) -> None:
         self.post_init_called = True
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py:110]: The file permissions in this commit appear to have been changed incorrectly, from 644 to 755.

**Comment by AichenF** [comment]
> 图片
> Is there an error in the data in the table? After enabling with torch.compile and Max-autotune, the total time increased from 53.3500s to 219.2028s, but the time per step decreased from 502ms to 190ms. Why is that?

Here, the total time comes from the log outputs, but our recent profiling shows that this time is not accurate because it does not include GPU execution time. From the profiling results, each step becomes faster after enabling torch.compile, but there incurs extra compilation overhead, so the overall time actually becomes longer.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 7]
```diff
@@ -151,6 +143,26 @@ def __init__(
...
+        if not hasattr(module, "forward"):
+            return module
+        try:
+            import torch._inductor.config as _inductor_cfg
+
+            _inductor_cfg.reorder_for_compute_comm_overlap = True
+        except ImportError:
+            pass
+        mode = os.environ.get("SGLANG_TORCH_COMPILE_MODE", "max-autotune-no-cudagraphs")
+        compiled_forward = torch.compile(getattr(module, "forward"), mode=mode)
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:167]: Add a log before this line is better? Such as `logger.info(f"Compiling transformer with mode: {mode}")`

**Review by BBuf — APPROVED** [review]
LGTM. In addition, it is recommended to focus primarily on the DenoisingStage section when reporting performance through the profiler, as this can help avoid some misunderstandings.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 8]
```diff
@@ -435,7 +444,7 @@ def _prepare_denoising_loop(self, batch: Req, server_args: ServerArgs):
 
         if batch.do_classifier_free_guidance:
             neg_cond_kwargs = self.prepare_extra_func_kwargs(
-                self.transformer.forward,
+                getattr(self.transformer, "forward", self.transformer),
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:621]: Can we add a test?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 9]
```diff
@@ -926,9 +935,29 @@ def prepare_extra_func_kwargs(self, func, kwargs) -> dict[str, Any]:
...
         """
         extra_step_kwargs = {}
+        accepted_params: set[str] = set()
+        # Try to inspect the callable directly
+        try:
+            accepted_params |= set(inspect.signature(func).parameters.keys())
+        except Exception:
+            pass
+        # If it's a module-like object, also try its forward
+        try:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Add a test too?

**Comment by BBuf** [comment]
Resolve conflicts.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 8 above)
Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:621]: like a ut for prepare_extra_func_kwargs?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 10]
```diff
@@ -333,6 +345,92 @@ def parallelism_type(self) -> StageParallelismType:
         # return StageParallelismType.CFG_PARALLEL if get_global_server_args().enable_cfg_parallel else StageParallelismType.REPLICATED
         return StageParallelismType.REPLICATED
 
+    def _prepare_denoising_loop(self, batch: Req, server_args: ServerArgs):
```
Comment (zcnrex) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: It seems there's already a `_prepare_denoising_loop()` function in this file.

File: python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py [hunk 11]
```diff
@@ -110,9 +110,7 @@ def post_init(self) -> None:
...
 
         self.initialize_pipeline(self.server_args)
-        if self.server_args.enable_torch_compile:
-            self.modules["transformer"] = torch.compile(self.modules["transformer"])
-            logger.info("Torch Compile enabled for DiT")
+        # NOTE: torch.compile is managed in the denoising stage to cover lazy-load paths.
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py]: nit: remove this

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 10 above)
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: @AichenF Please see this comment.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 8 above)
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:621]: Never mind, no need to add it.

**Comment by AichenF** [comment]
> I noticed that there's also a torch.compile at [this location](https://github.com/sgl-project/sglang/blob/main/python/sglang/multimodal_gen/runtime/pipelines_core/composed_pipeline_base.py#L112-L115). What is the relationship between this torch.compile and the one in the current `denoising.py`? Could there be a case of redundant compilation? cc @mickqian

this compile is redundant, deleting it will not affect the compile process, i have deleted it.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 10 above)
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: @AichenF here

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 10 above)
Comment (AichenF) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py]: Thanks for the heads‑up, this really was a duplicate.

**Review by BBuf — APPROVED** [review]
LGTM.
